core/src/commonIntegrationTest/kotlin/com/powersync/DatabaseTest.kt

```diff
@@ -114,6 +230,71 @@ class DatabaseTest {
             }
         }
 
+    @Test
+    fun testClosingReadPool() =
+        runTest {
+            val pausedLock = CompletableDeferred<Unit>()
+            val inLock = CompletableDeferred<Unit>()
+            // Request a lock
+            val lockJob =
+                async {
+                    database.readLock {
+                        inLock.complete(Unit)
+                        runBlocking {
+                            pausedLock.await()
+                        }
+                    }
+                }
+
+            // Wait for the lock to be active
+            inLock.await()
+
+            // Close the database. This should close the read pool
+            // The pool should wait for jobs to complete before closing
+            val closeJob =
+                async {
+                    database.close()
+                }
+
+            // Wait a little for testing
+            // Spawns in a different context for the delay to actually take affect
+            async { withContext(Dispatchers.Default) { delay(500) } }.await()
```

Comment:
Not a huge fan of this, but I can't see a much better way either. We could add this to `ConnectionPool`:

```Kotlin
    @OptIn(DelicateCoroutinesApi::class)
    internal val closeRequested: Boolean get() = available.isClosedForReceive
```

But when we'd still have to expose the pool on `InternalDatabaseImpl` and the database impl on `PowerSyncDatabaseImpl` as `internal` fields to access them in this test (to poll on `closeRequested` becoming `true`).

Reply:
Discussed offline. This delay here is to ensure that the database did not close immediately after the `close` method was called since there is an active readLock open. The close method should wait for the lock to complete.  
We currently only mark the database as `closed=true` after the database is actually closed. During this state some operations might succeed while others will fail due to the underlaying database being closed.